habitat-lab/habitat/sims/habitat_simulator/sim_utilities.py

```diff
@@ -754,6 +754,74 @@ def within(
     return containment_ids
 
 
+def ontop(
+    sim: habitat_sim.Simulator,
+    objectA: Union[
+        habitat_sim.physics.ManagedRigidObject,
+        habitat_sim.physics.ManagedArticulatedObject,
+        int,
+    ],
+    do_collision_detection: bool = True,
+    vertical_normal_error_threshold: float = 0.75,
+) -> List[int]:
+    """
+    Get a list of all object ids or objects that are "ontop" of a particular objectA.
+    Concretely, 'ontop' is defined as: contact points between objectA and objectB have vertical normals "upward" relative to objectA.
+    This function uses collision points to determine which objects are resting on or contacting the surface of objectA.
+
+    :param sim: The Simulator instance.
+    :param objectA: The ManagedRigidObject or object id for which to query the 'ontop' set.
+    :param do_collision_detection: If True, a fresh discrete collision detection is run before the contact point query. Pass False to skip if a recent sim step or pre-process has run a collision detection pass on the current state.
+    :param vertical_normal_error_threshold: The allowed error in normal alignment for a contact point to be considered "vertical" for this check. Functionally, if dot(contact normal, Y) <= threshold, the contact is ignored.
+
+    :return: a list of integer object_ids for the set of objects "ontop" of objectA.
+    """
+
+    link_id = None
+    if isinstance(objectA, int):
+        subject_object = get_obj_from_id(sim, objectA)
+        if subject_object is None:
+            raise AssertionError(f"The passed object_id {objectA} is invalid.")
+        if subject_object.object_id != objectA:
+            # objectA is a link
+            link_id = subject_object.link_object_ids[objectA]
+        objectA = subject_object
+
+    if do_collision_detection:
+        sim.perform_discrete_collision_detection()
+
+    yup = mn.Vector3(0.0, 1.0, 0.0)
+
+    ontop_object_ids = []
+    for cp in sim.get_physics_contact_points():
+        contacting_obj_id = None
+        obj_is_b = False
+        if cp.object_id_a == objectA.object_id and (
+            link_id is None or link_id == cp.link_id_a
+        ):
+            contacting_obj_id = cp.object_id_b
+        elif cp.object_id_b == objectA.object_id and (
+            link_id is None or link_id == cp.link_id_b
+        ):
+            contacting_obj_id = cp.object_id_a
+            obj_is_b = True
```

Comment:
I'm probably misunderstanding what's going on here.

Wouldn't we want to exclude the object itself from the set of objects on top of it? Would a `continue` be better suited here here?

Reply:
Yeah, possibly misunderstanding. For clarity, we don't know if our object is the first or second in the contact pair. This statement just figures out which one it is if either and prepares to invert the normal if necessary.